### PR TITLE
Optimize ClickHouse schema and enhance data processing

### PR DESCRIPTION
**Pull request by iuwqyir**
### TL;DR

Updated ClickHouse schema and data handling for blocks, transactions, logs, and traces.

### What changed?

- Added `total_difficulty` field to blocks table
- Updated transactions table:
  - Renamed `input` to `data`
  - Added `r`, `s`, `v`, and `access_list` fields
- Modified traces table structure:
  - Removed `id` field
  - Added `author`, `reward_type`, and `refund_address` fields
- Adjusted logs table indexing
- Updated ClickHouseConnector methods to reflect schema changes
- Modified MemoryConnector's trace handling

### How to test?

1. Apply the schema changes to your ClickHouse database
2. Update any existing queries or data insertion methods to match the new schema
3. Verify that blocks, transactions, logs, and traces are correctly inserted and retrieved with the new fields
4. Test the updated MemoryConnector with trace operations

### CODE REVIEW

**Comment by iuwqyir** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/thirdweb-dev/indexer/47?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#47** <a href="https://app.graphite.dev/github/pr/thirdweb-dev/indexer/47?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#46** <a href="https://app.graphite.dev/github/pr/thirdweb-dev/indexer/46?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#45** <a href="https://app.graphite.dev/github/pr/thirdweb-dev/indexer/45?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#42** <a href="https://app.graphite.dev/github/pr/thirdweb-dev/indexer/42?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#41** <a href="https://app.graphite.dev/github/pr/thirdweb-dev/indexer/41?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @iuwqyir and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: internal/tools/clickhouse_create_blocks_table.sql [hunk 1]
```diff
@@ -25,5 +26,5 @@ CREATE TABLE base.blocks (
     INDEX idx_timestamp timestamp TYPE minmax GRANULARITY 1,
     INDEX idx_number number TYPE minmax GRANULARITY 1,
 ) ENGINE = ReplacingMergeTree(insert_timestamp, is_deleted)
-ORDER BY (chain_id, hash) PRIMARY KEY (chain_id, hash)
+ORDER BY (chain_id, hash)
```
Comment (AmineAfia) [inline internal/tools/clickhouse_create_blocks_table.sql:29]: why?

File: internal/tools/clickhouse_create_logs_table.sql [hunk 2]
```diff
@@ -8,18 +8,17 @@ CREATE TABLE base.logs (
...
-    INDEX topic2_idx topic_2 TYPE bloom_filter GRANULARITY 1,
-    INDEX topic3_idx topic_3 TYPE bloom_filter GRANULARITY 1,
+    INDEX idx_block_timestamp block_timestamp TYPE minmax GRANULARITY 1,
+    INDEX idx_block_number block_number TYPE minmax GRANULARITY 1,
+    INDEX idx_block_hash block_hash TYPE bloom_filter GRANULARITY 1,
+    INDEX idx_address address TYPE bloom_filter GRANULARITY 1,
+    INDEX idx_topic0 topic_0 TYPE bloom_filter GRANULARITY 1,
 ) ENGINE = SharedReplacingMergeTree(insert_timestamp, is_deleted)
-ORDER BY (block_number, transaction_hash, log_index)
+ORDER BY (chain_id, transaction_hash, log_index, block_hash)
```
Comment (AmineAfia) [inline internal/tools/clickhouse_create_logs_table.sql:23]: why not blocknumber?

File: internal/tools/clickhouse_create_traces_table.sql [hunk 3]
```diff
@@ -1,27 +1,33 @@
...
+    INDEX idx_block_timestamp block_timestamp TYPE minmax GRANULARITY 1,
+    INDEX idx_block_number block_number TYPE minmax GRANULARITY 1,
+    INDEX idx_block_hash block_hash TYPE bloom_filter GRANULARITY 1,
+    INDEX idx_from_address from_address TYPE bloom_filter GRANULARITY 1,
+    INDEX idx_to_address to_address TYPE bloom_filter GRANULARITY 1,
+    INDEX idx_type type TYPE bloom_filter GRANULARITY 1,
 ) ENGINE = ReplacingMergeTree(insert_timestamp, is_deleted)
-ORDER BY (block_number, id) SETTINGS allow_experimental_replacing_merge_with_cleanup = 1;
\ No newline at end of file
+ORDER BY (chain_id, transaction_hash, trace_address, block_hash)
```
Comment (AmineAfia) [inline internal/tools/clickhouse_create_traces_table.sql:32]: why block_hash and not block_number?

**Review by AmineAfia — COMMENTED** [review]
before we merge, I have a question regarding table ordering

File: internal/tools/clickhouse_create_blocks_table.sql (hunk 1 above)
Comment (iuwqyir) [reply internal/tools/clickhouse_create_blocks_table.sql:29]: We do not need a primary key, we have the sort order and indexes defined

File: internal/tools/clickhouse_create_logs_table.sql (hunk 2 above)
Comment (iuwqyir) [reply internal/tools/clickhouse_create_logs_table.sql:23]: we have a block number index. Makes sense to setup tables by hash by default

File: internal/tools/clickhouse_create_traces_table.sql (hunk 3 above)
Comment (iuwqyir) [reply internal/tools/clickhouse_create_traces_table.sql:32]: same as before. It will persist data even in case of reorgs

**Review by AmineAfia — APPROVED** [review]

**Comment by iuwqyir** [comment]
### Merge activity

* **Sep 25, 6:48 PM EDT**: @iuwqyir started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/thirdweb-dev/indexer/47).
